### DISCUSSION TITLE
Handling thousands of draggable elements?

### DISCUSSION THREAD

**Discussion by CrageJ** [clauderic/dnd-kit#1482] [General]
I'm working on a project which will render thousands of draggable grid elements. By itself the library could not handle more than a few hundred elements, and using lazy rendering via suspense did not help. Would there be any further optimizations possible for this use case? Are there any other libraries, including non-react libraries better suited for this task?

I did my experiments [here](https://codesandbox.io/p/sandbox/dndkit-grid-with-drag-handle-example-forked-wltv2d?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cm0e3zw5q00062868vggwcwp7%2522%252C%2522sizes%2522%253A%255B100%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cm0e3zw5q00022868fpdd4kzp%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cm0e3zw5q00032868ryi2b7mt%2522%257D%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cm0e3zw5q000528683jifbgxw%2522%257D%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cm0e3zw5q00022868fpdd4kzp%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm0e3zw5p00012868w2whcim4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522cm0e42apk002q2868k2eqgj3c%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A53%252C%2522startColumn%2522%253A9%252C%2522endLineNumber%2522%253A53%252C%2522endColumn%2522%253A9%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522cm0e3zw5q00022868fpdd4kzp%2522%252C%2522activeTabId%2522%253A%2522cm0e42apk002q2868k2eqgj3c%2522%257D%252C%2522cm0e3zw5q000528683jifbgxw%2522%253A%257B%2522id%2522%253A%2522cm0e3zw5q000528683jifbgxw%2522%252C%2522activeTabId%2522%253A%2522cm0e404dv001c2868vzndpv3z%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm0e3zw5q00042868f7h3kv23%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522id%2522%253A%2522cm0e404dv001c2868vzndpv3z%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522cm0e3zw5q00032868ryi2b7mt%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cm0e3zw5q00032868ryi2b7mt%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D). This was generated by chatGPT as I have not used react before.

**Comment by LordZardeck** [comment]
We had a similar issue, tons of VERY complex components, getting constantly re-rendered on dragmove, ended up solving it by creating a component that used the dnd hooks, then passed only the data the component would need to a memoized slot. This way re-rendering stopped at the hook unless something changed that the component cared about. For example:

```ts
import { memo } from 'react'
import { Slot } from '@radix-ui/react-slot'

export const MemoizedSlot = memo(Slot)
```

```ts
import type { ForwardedRef, PropsWithChildren } from 'react'
import { forwardRef, useMemo } from 'react'
import { MemoizedSlot } from '@core/ui/utils'
import { useDroppable } from '@dnd-kit/core'
import { mergeRefs } from 'react-merge-refs'

export type DroppableProps = PropsWithChildren<
	Omit<Parameters<typeof useDroppable>[0], 'id'> & {
		id: string
	}
>

export const Droppable = forwardRef(function Droppable<R extends HTMLElement>(
	props: DroppableProps,
	ref: ForwardedRef<R>,
) {
	const { isOver, setNodeRef } = useDroppable(props)
	const memoizedRef = useMemo(() => mergeRefs([ref, setNodeRef]), [ref, setNodeRef])

	return (
		<MemoizedSlot
			{...props}
			ref={memoizedRef}
			data-drag-over={isOver}
		/>
	)
})
```
